bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java

```diff
@@ -262,6 +270,47 @@ public void testCanHandleTopLevelGenericIfHasCodec() {
                 "{'myGenericField': {'$numberLong': '5'}}");
     }
 
+    @Test
+    public void testCustomRegisteredPropertyCodecWithValue() {
+        CustomPropertyCodecOptionalModel model = new CustomPropertyCodecOptionalModel(Optional.of("foo"));
+        roundTrip(getPojoCodecProviderBuilder(CustomPropertyCodecOptionalModel.class).register(new OptionalPropertyCodecProviderForTest()),
+                model, "{'optionalField': 'foo'}");
+    }
+
+    @Test
+    public void testCustomRegisteredPropertyCodecOmittedValue() {
+        CustomPropertyCodecOptionalModel model = new CustomPropertyCodecOptionalModel(Optional.<String>empty());
+        roundTrip(getPojoCodecProviderBuilder(CustomPropertyCodecOptionalModel.class).register(new OptionalPropertyCodecProviderForTest()),
+                model, "{'optionalField': null}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testEncodingInvalidMapModel() {
+        encodesTo(getPojoCodecProviderBuilder(InvalidMapModel.class), getInvalidMapModel(), "{'invalidMap': {'1': 1, '2': 2}}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testDecodingInvalidMapModel() {
+        decodingShouldFail(getCodec(InvalidMapModel.class), "{'invalidMap': {'1': 1, '2': 2}}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testEncodingInvalidCollectionModel() {
+        encodesTo(getPojoCodecProviderBuilder(InvalidCollectionModel.class), new InvalidCollectionModel(asList(1, 2, 3)),
+                "{collectionField: [1, 2, 3]}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testDecodingInvalidCollectionModel() {
+        decodingShouldFail(getCodec(InvalidCollectionModel.class), "{collectionField: [1, 2, 3]}");
+    }
+
+    @Test
+    public void testInvalidMapModelWithCustomPropertyCodecProvider() {
+        encodesTo(getPojoCodecProviderBuilder(InvalidMapModel.class).register(new InvalidMapPropertyCodecProvider()), getInvalidMapModel(),
```

Comment:
Custom `PropertyCodecProviders` also allow us to handle previously invalid Map implementations.